src/Servers/HttpSys/src/AsyncAcceptContext.cs

```diff
@@ -2,121 +2,133 @@
+        internal ValueTask<RequestContext> AcceptAsync()
+        {
+            _mrvts.Reset();
+
+            AllocateNativeRequest();
+
+            uint statusCode = QueueBeginGetContext();
+            if (statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS &&
+                statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_IO_PENDING)
+            {
+                // some other bad error, possible(?) return values are:
+                // ERROR_INVALID_HANDLE, ERROR_INSUFFICIENT_BUFFER, ERROR_OPERATION_ABORTED
+                return ValueTask.FromException<RequestContext>(new HttpSysException((int)statusCode));
+            }
+
+            return new ValueTask<RequestContext>(this, _mrvts.Version);
+        }
+
         private static void IOCompleted(AsyncAcceptContext asyncContext, uint errorCode, uint numBytes)
         {
             bool complete = false;
+
             try
             {
                 if (errorCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS &&
                     errorCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_MORE_DATA)
                 {
-                    asyncContext.TrySetException(new HttpSysException((int)errorCode));
-                    complete = true;
+                    asyncContext._mrvts.SetException(new HttpSysException((int)errorCode));
+                    return;
                 }
-                else
+
+                HttpSysListener server = asyncContext.Server;
+                if (errorCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS)
                 {
-                    HttpSysListener server = asyncContext.Server;
-                    if (errorCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS)
+                    // at this point we have received an unmanaged HTTP_REQUEST and memoryBlob
+                    // points to it we need to hook up our authentication handling code here.
+                    try
                     {
-                        // at this point we have received an unmanaged HTTP_REQUEST and memoryBlob
-                        // points to it we need to hook up our authentication handling code here.
-                        try
-                        {
-                            if (server.ValidateRequest(asyncContext._nativeRequestContext) && server.ValidateAuth(asyncContext._nativeRequestContext))
-                            {
-                                RequestContext requestContext = new RequestContext(server, asyncContext._nativeRequestContext);
-                                asyncContext.TrySetResult(requestContext);
-                                complete = true;
-                            }
-                        }
-                        catch (Exception)
-                        {
-                            server.SendError(asyncContext._nativeRequestContext.RequestId, StatusCodes.Status400BadRequest);
```

Comment:
I was wondering if you'd catch it. I wanted to see if we thought it was important. I'll move it to the MessagePump is the accept fails.

Reply:
Exceptions should bubble to the caller

Reply:
> I'll move it to the MessagePump is the accept fails.

How? The message pump doesn't have access to the request if the Accept failed part way through.



> Exceptions should bubble to the caller

It can, but the error response still needs to be sent in order to prevent a native memory leak.

Reply:
I might be missing something but it seems the only thing passed from the request was the ID. I would make a new exception type that stored the ID and use that to write the response.